### PR TITLE
[SPARK-45093][CONNECT][PYTHON] Error reporting for addArtifacts query

### PR DESCRIPTION
**Pull request by cdkrot**
### What changes were proposed in this pull request?

Add error logging into `addArtifact`  (see example in "How this is tested"). The logging code is moved into separate file to avoid circular dependency.

### Why are the changes needed?

Currently, in case `addArtifact` is executed with the file which doesn't exist, the user gets cryptic error

```grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Exception iterating requests!"
        debug_error_string = "None"
>
```

Which is hard to debug without deep digging into the subject.

This happens because addArtifact is implemented as client-side streaming and the actual error happens during grpc consuming iterator generating requests. Unfortunately grpc doesn't print any debug information for user to understand the problem.

### Does this PR introduce _any_ user-facing change?

Additional logging which is opt-in same way as before with `SPARK_CONNECT_LOG_LEVEL` environment variable.

### How was this patch tested?

```
>>> s.addArtifact("XYZ", file=True)
[New:] 2023-09-15 17:06:40,078 11789 ERROR _create_requests Failed to execute addArtifact: [Errno 2] No such file or directory: '/Users/alice.sayutina/apache_spark/python/XYZ'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alice.sayutina/apache_spark/python/pyspark/sql/connect/session.py", line 743, in addArtifacts
    self._client.add_artifacts(*path, pyfile=pyfile, archive=archive, file=file)

[....]

  File "/Users/alice.sayutina/oss-venv/lib/python3.11/site-packages/grpc/_channel.py", line 910, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "Exception iterating requests!"
        debug_error_string = "None"
>

```

### CODE REVIEW

**Comment by cdkrot** [comment]
cc @HyukjinKwon, @nija-at

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/client/logging.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import logging
+import os
+from typing import Optional
+
+__all__ = [
+    "logger",
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/logging.py:6]:
```suggestion
```

I suspect `logger` isn't an API?

File: python/pyspark/sql/connect/client/artifact.py [hunk 2]
```diff
@@ -271,6 +276,7 @@ def add_artifacts(self, *path: str, pyfile: bool, archive: bool, file: bool) ->
         requests: Iterator[proto.AddArtifactsRequest] = self._create_requests(
             *path, pyfile=pyfile, archive=archive, file=file
         )
+
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/artifact.py:279]:
```suggestion
```

Since we're here, I would remove the unrelated change.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/client/artifact.py [hunk 3]
```diff
@@ -243,11 +244,15 @@ def _create_requests(
...
-                *(self._parse_artifacts(p, pyfile=pyfile, archive=archive, file=file) for p in path)
+        try:
+            yield from self._add_artifacts(
+                chain(
+                    *(self._parse_artifacts(p, pyfile=pyfile, archive=archive, file=file) for p in path)
+                )
             )
-        )
+        except Exception as e:
+            logger.error(f"Failed to submit addArtifact request: {e}")
```
Comment (nija-at) [inline python/pyspark/sql/connect/client/artifact.py:254]:
```suggestion
            logger.error(f"Failed to submit addArtifacts request: {e}")
```

**Review by nija-at — APPROVED** [review]

File: python/pyspark/sql/connect/client/logging.py (hunk 1 above)
Comment (cdkrot) [reply python/pyspark/sql/connect/client/logging.py:6]: logger isn't an API, yeah. I didn't import it in __init__, but probably best to drop it here too.

File: python/pyspark/sql/connect/client/logging.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import logging
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/logging.py:1]: Seems like linter complains that there's no license header :-).

File: python/pyspark/sql/connect/client/artifact.py [hunk 5]
```diff
@@ -243,11 +244,15 @@ def _create_requests(
...
-                *(self._parse_artifacts(p, pyfile=pyfile, archive=archive, file=file) for p in path)
+        try:
+            yield from self._add_artifacts(
+                chain(
+                    *(self._parse_artifacts(p, pyfile=pyfile, archive=archive, file=file) for p in path)
+                )
             )
-        )
+        except Exception as e:
+            logger.error(f"Failed to submit addArtifacts request: {e}")
```
Comment (allisonwang-db) [inline python/pyspark/sql/connect/client/artifact.py:254]: I think we should show this error message regardless of the `SPARK_CONNECT_LOG_LEVEL` setting to make it more debuggable. Can we raise a more user-friendly error below?

File: python/pyspark/sql/connect/client/logging.py (hunk 4 above)
Comment (cdkrot) [reply python/pyspark/sql/connect/client/logging.py:1]: Thanks :). Forgot about that. Done.

File: python/pyspark/sql/connect/client/artifact.py (hunk 5 above)
Comment (cdkrot) [reply python/pyspark/sql/connect/client/artifact.py:254]: The problem is not that we don't raise error below, but that it will be intercepted by `grpc`, and it doesn't log what was thrown (I would argue that it definitely should, but.).

**Comment by cdkrot** [comment]
From PR script:

```Run PYTHON_EXECUTABLE=python3.9 ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting black test...
black checks failed:
Oh no! 💥 💔 💥 The required version `23.9.1` does not match the running version `22.6.0`!
Please run 'dev/reformat-python' script.
```

Should I do something about it or is it a known problem?

File: python/pyspark/sql/connect/client/artifact.py (hunk 5 above)
Comment (allisonwang-db) [reply python/pyspark/sql/connect/client/artifact.py:254]: In core.py, there is a method to handle errors from the server: `def _handle_error(error)`. It's able to handle various GRPC errors and also custom errors. Can we have something similar for add artifacts?

**Comment by HyukjinKwon** [comment]
Ah, the master branch and your branch should be synced to the latest master brnach. Black was latenly upgraded.

**Comment by cdkrot** [comment]
Updated fork's master

File: python/pyspark/sql/connect/client/artifact.py (hunk 5 above)
Comment (allisonwang-db) [reply python/pyspark/sql/connect/client/artifact.py:254]: Could you be more specific on what `Exception` can be thrown here? From your PR description, it looks like a GPRC error.

Comment (allisonwang-db) [reply python/pyspark/sql/connect/client/artifact.py:254]: What's the behavior after your change when the SPARK_CONNECT logger is enabled?

Comment (allisonwang-db) [reply python/pyspark/sql/connect/client/artifact.py:254]: I just think we shouldn't leave users this error message, when they don't enable the spark connect logger:
```
        status = StatusCode.UNKNOWN
        details = "Exception iterating requests!"
        debug_error_string = "None"
```

Comment (cdkrot) [reply python/pyspark/sql/connect/client/artifact.py:254]: In my example I'm intercepting FileNotFound error

Comment (cdkrot) [reply python/pyspark/sql/connect/client/artifact.py:254]: > What's the behavior after your change when the SPARK_CONNECT logger is enabled?
described in pr

**Comment by HyukjinKwon** [comment]
Mind running `dev/reformat-python` one more last time? Seems like there's a diff between old and new black versions.

**Comment by cdkrot** [comment]
Done, thanks!

**Comment by HyukjinKwon** [comment]
Merged to master.
